src/reports/models/reportsManager.js

```diff
@@ -102,4 +130,72 @@ function generateGraphanaUrl(report, grafanaUrl) {
         const grafanaReportUrl = encodeURI(grafanaUrl + `&var-Name=${report.test_name}&from=${new Date(report.start_time).getTime()}${endTimeGrafanafaQuery}`);
         return grafanaReportUrl;
     }
-}
\ No newline at end of file
+}
+
+function calculateReportStatus(report) {
+    const subscribersStages = getListOfSubscribersStages(report);
+    const uniqueSubscribersStages = _.uniq(subscribersStages);
+
+    const delayedTimeInMs = Math.max(report.duration * 0.01, MINIMUM_WAIT_FOR_DELAYED_REPORT_UPDATE_MS);
+    const reportDurationMs = report.duration * 1000;
+    const reportStartTimeMs = new Date(report.start_time).getTime();
+
+    if (allSubscribersFinished(uniqueSubscribersStages)) {
+        return constants.REPORT_FINISHED_STATUS;
+    } else if (Date.now() >= reportStartTimeMs + reportDurationMs + delayedTimeInMs) {
+        if (uniqueSubscribersStages.includes(constants.SUBSCRIBER_DONE_STAGE)) {
+            return constants.REPORT_PARTIALLY_FINISHED_STATUS;
+        } else {
+            return constants.REPORT_FAILED_STATUS;
+        }
+    } else if (allSubscribersInitializing(uniqueSubscribersStages)) {
+        return constants.REPORT_INITIALIZING_STATUS;
+    } else if (allSubscribersStarted(uniqueSubscribersStages)) {
+        return constants.REPORT_STARTED_STATUS;
+    } else if (allSubscribersAborted(uniqueSubscribersStages)) {
+        return constants.REPORT_ABORTED_STATUS;
+    } else if (allSubscribersFailed(uniqueSubscribersStages)) {
+        return constants.REPORT_FAILED_STATUS;
+    } else {
+        return calculateDynamicReportStatus(report, uniqueSubscribersStages);
+    }
+}
+
+function calculateDynamicReportStatus(report, uniqueSubscribersStages) {
+    if (uniqueSubscribersStages.includes(constants.SUBSCRIBER_DONE_STAGE)) {
+        return constants.REPORT_PARTIALLY_FINISHED_STATUS;
+    } else if (uniqueSubscribersStages.includes(constants.SUBSCRIBER_INTERMEDIATE_STAGE)) {
+        return constants.REPORT_IN_PROGRESS_STATUS;
+    } else if (uniqueSubscribersStages.includes(constants.SUBSCRIBER_STARTED_STAGE)) {
+        return constants.REPORT_STARTED_STATUS;
+    } else if (uniqueSubscribersStages.includes(constants.SUBSCRIBER_INITIALIZING_STAGE)) {
+        return constants.REPORT_INITIALIZING_STATUS;
+    } else {
+        return constants.REPORT_FAILED_STATUS;
+    }
+}
+
+function getListOfSubscribersStages (report) {
+    const runnerStates = report.subscribers.map((subscriber) => subscriber.stage);
+    return runnerStates;
+}
+
+function allSubscribersFinished (subscriberStages) {
```

Comment:
same function different param 
can be refactored

Reply:
done